### PR TITLE
network-libp2p: Migrate AutoNAT v1 protocol to v2

### PR DESCRIPTION
**Pull request by Eligioo**
## What's in this pull request?
Transition from AutoNAT v1 protocol to v2
Main differences:
- `autonat v1` allowed testing reachability for the node. `autonat v2` allows testing reachability for an individual address.
- `autonat v2` provides a mechanism for nodes to verify whether the peer actually successfully dialled an address.
- `autonat v2` provides a mechanism for nodes to dial an IP address different from the requesting node's observed IP address without risking amplification attacks. `autonat v1` disallowed such dials to prevent amplification attacks.

This PR should contribute improving the issues stated by #276 

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

File: network-libp2p/src/autonat.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
+use std::collections::{HashMap, HashSet};
+
+use libp2p::Multiaddr;
+
+#[derive(Clone, Debug, Default, PartialEq)]
```
Comment (nibhar) [inline network-libp2p/src/autonat.rs]: This could probably get `Copy`, though it does not do much here.

File: network-libp2p/src/autonat.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    pub fn set_listen_address_nat_status(&mut self, address: Multiaddr, nat_status: NatStatus) {
+        if let Some(status) = self.listen_address_status.get_mut(&address) {
+            *status = nat_status.clone();
+
+            if nat_status == NatStatus::Public {
+                self.confirmed_addresses.insert(address);
+            } else {
+                self.confirmed_addresses.remove(&address);
+            }
```
Comment (nibhar) [inline network-libp2p/src/autonat.rs]: Swap around to avoid the `.clone()`.
```suggestion
            if nat_status == NatStatus::Public {
                self.confirmed_addresses.insert(address);
            } else {
                self.confirmed_addresses.remove(&address);
            }

            *status = nat_status;
```

File: network-libp2p/tests/network.rs [hunk 3]
```diff
@@ -318,7 +317,7 @@ async fn create_network_with_n_peers(n_peers: usize) -> Vec<Network> {
 
 #[test(tokio::test)]
 async fn connections_stress_and_reconnect() {
-    let peers: usize = 10;
+    let peers: usize = 5;
```
Comment (nibhar) [inline network-libp2p/tests/network.rs:320]: Why this change?

**Review by nibhar — APPROVED** [review]
LGTM, but i have limited knowledge of the inner workings of libp2p here. Probably a good idea to have another reviewer as well.

File: network-libp2p/src/autonat.rs (hunk 1 above)
Comment (Eligioo) [reply network-libp2p/src/autonat.rs]: Is now a `Copy` 👍

File: network-libp2p/src/autonat.rs (hunk 2 above)
Comment (Eligioo) [reply network-libp2p/src/autonat.rs]: Swap not necessary since it's now a `Copy` 👍

File: network-libp2p/tests/network.rs (hunk 3 above)
Comment (Eligioo) [reply network-libp2p/tests/network.rs:320]: This test is flaky and with the old number of peers, the network in the CI would progress too slow as there is more going on during network bootstrapping (namely a multi roundtrip process for the Autonat probes per listen adress).

Peers start to disconnect after 30 seconds thus assertions don't hold anymore. With 5 peers it seems the CI is able to keep up.

Comment (nibhar) [reply network-libp2p/tests/network.rs:320]: Okay. We should maybe change the test setup then eventually.

File: network-libp2p/src/autonat.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    Public,
+    Private,
+    #[default]
+    Unknown,
+}
+
+#[derive(Default)]
+pub(crate) struct NatState {
+    confirmed_addresses: HashSet<Multiaddr>,
+    listen_address_status: HashMap<Multiaddr, NatStatus>,
```
Comment (jsdanielh) [inline network-libp2p/src/autonat.rs]: Why listen? Maybe only: `addresses_status`?

File: network-libp2p/src/autonat.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    status: NatStatus,
+}
+
+impl NatState {
+    pub fn add_listen_address(&mut self, address: Multiaddr) {
```
Comment (jsdanielh) [inline network-libp2p/src/autonat.rs]: Maybe only `add_address`?

File: network-libp2p/src/autonat.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
+use std::collections::{HashMap, HashSet};
+
+use libp2p::Multiaddr;
+
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+pub(crate) enum NatStatus {
```
Comment (jsdanielh) [inline network-libp2p/src/autonat.rs:7]: Could you add some `rustdoc`?

File: network-libp2p/src/autonat.rs (hunk 5 above)
Comment (jsdanielh) [inline network-libp2p/src/autonat.rs]: Could you add some `rustdoc`?

File: network-libp2p/src/autonat.rs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    status: NatStatus,
+}
+
+impl NatState {
+    pub fn add_listen_address(&mut self, address: Multiaddr) {
+        self.listen_address_status
+            .insert(address, NatStatus::Unknown);
+    }
+
+    pub fn remove_listen_address(&mut self, address: &Multiaddr) {
```
Comment (jsdanielh) [inline network-libp2p/src/autonat.rs]: Could you add some `rustdoc`?

File: network-libp2p/src/autonat.rs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        if let Some(status) = self.listen_address_status.get_mut(&address) {
+            *status = nat_status;
+
+            if nat_status == NatStatus::Public {
+                self.confirmed_addresses.insert(address);
+            } else {
+                self.confirmed_addresses.remove(&address);
+            }
+            self.update_state();
+        }
```
Comment (jsdanielh) [inline network-libp2p/src/autonat.rs]:
```suggestion
        let Some(status) = self.listen_address_status.get_mut(&address) else {return;};
        *status = nat_status;

        if nat_status == NatStatus::Public {
            self.confirmed_addresses.insert(address);
        } else {
            self.confirmed_addresses.remove(&address);
        }
        self.update_state();
}
```

File: network-libp2p/src/autonat.rs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    pub fn add_confirmed_address(&mut self, address: Multiaddr) {
+        if let Some(address_status) = self.listen_address_status.get_mut(&address) {
```
Comment (jsdanielh) [inline network-libp2p/src/autonat.rs]: To me, we should add it to `listen_address` if it doesn't exist and add them to `confirmed_addresses` regardless

File: network-libp2p/src/autonat.rs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            *address_status = NatStatus::Public;
+
+            self.confirmed_addresses.insert(address);
+            self.update_state();
+        }
+    }
+
+    pub fn remove_confirmed_address(&mut self, address: &Multiaddr) {
+        if let Some(address_status) = self.listen_address_status.get_mut(address) {
```
Comment (jsdanielh) [inline network-libp2p/src/autonat.rs]: To me, we should add it to `listen_address` if it doesn't exist and remove it from `confirmed_addresses` regardless

File: network-libp2p/src/network_types.rs [hunk 11]
```diff
@@ -238,6 +239,8 @@ pub(crate) struct TaskState {
     pub(crate) dht_bootstrap_state: DhtBootStrapState,
     /// DHT (kad) is in server mode
     pub(crate) dht_server_mode: bool,
+    /// The NAT status of the local peer
+    pub(crate) nat_status: NatState,
```
Comment (jsdanielh) [inline network-libp2p/src/network_types.rs:243]: Maybe I'm missing something here, but why do we want this `nat_status`? I don't see any reader of this

**Comment by jsdanielh** [comment]
@Eligioo I guess that with autonat v2, [this](https://github.com/nimiq/core-rs-albatross/blob/albatross/lib/src/config/config_file/client.example.toml#L37) is no longer needed.

File: network-libp2p/src/autonat.rs (hunk 10 above)
Comment (Eligioo) [reply network-libp2p/src/autonat.rs]: Changed it to `self.listen_address_status.entry.or_default()` so it will always exists in `listen_address`

File: network-libp2p/src/autonat.rs (hunk 8 above)
Comment (Eligioo) [reply network-libp2p/src/autonat.rs]: There will always be an entry now

File: network-libp2p/src/network_types.rs (hunk 11 above)
Comment (Eligioo) [reply network-libp2p/src/network_types.rs:243]: It is necessary to keep a state of the all the addresses in order to determine the final NAT state. With `autonatv1`, it would give you a NAT of `Public` or `Private` eventually but v2 only gives you the results of the probes. Thus you have to keep track of everything yourself.

Comment (jsdanielh) [reply network-libp2p/src/network_types.rs:243]: But who's reading this status? I am only seeing some logging related to this

**Review by jsdanielh — APPROVED** [review]

File: network-libp2p/src/network_types.rs (hunk 11 above)
Comment (jsdanielh) [reply network-libp2p/src/network_types.rs:243]: We discussed it internally and we decided to keep it for `Nat` status visibility.
